spring-cloud-gcp-starters/spring-cloud-gcp-starter-pubsub/src/main/java/org/springframework/cloud/gcp/pubsub/autoconfig/GcpPubsubAutoConfiguration.java

```diff
@@ -0,0 +1,111 @@
+/*
+ *  Copyright 2017 original author or authors.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package org.springframework.cloud.gcp.pubsub.autoconfig;
+
+import java.util.concurrent.Executors;
+
+import com.google.api.gax.grpc.ChannelProvider;
+import com.google.api.gax.grpc.ExecutorProvider;
+import com.google.api.gax.grpc.FixedExecutorProvider;
+import com.google.auth.oauth2.GoogleCredentials;
+import com.google.cloud.pubsub.v1.SubscriptionAdminSettings;
+import com.google.cloud.pubsub.v1.TopicAdminSettings;
+
+import org.springframework.beans.factory.annotation.Qualifier;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.boot.autoconfigure.AutoConfigureAfter;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
+import org.springframework.cloud.gcp.core.GcpProperties;
+import org.springframework.cloud.gcp.core.autoconfig.GcpContextAutoConfiguration;
+import org.springframework.cloud.gcp.pubsub.core.PubSubTemplate;
+import org.springframework.cloud.gcp.pubsub.support.DefaultPublisherFactory;
+import org.springframework.cloud.gcp.pubsub.support.DefaultSubscriberFactory;
+import org.springframework.cloud.gcp.pubsub.support.PublisherFactory;
+import org.springframework.cloud.gcp.pubsub.support.SubscriberFactory;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+
+/**
+ * @author João André Martins
+ */
+@Configuration
+@AutoConfigureAfter(GcpContextAutoConfiguration.class)
+public class GcpPubsubAutoConfiguration {
+
+	public static final String DEFAULT_SOURCE_NAME = "spring";
+
+	@Value("${spring.cloud.gcp.pubsub.subscriber.executorThreads:4}")
+	private int subscriberExecutorThreads;
+
+	@Value("${spring.cloud.gcp.pubsub.publisher.executorThreads:4}")
+	private int publisherExecutorThreads;
+
+	@Bean
+	public ExecutorProvider publisherExecutorProvider() {
```

Comment:
The `@ConditionalOnMissingBean` must be here anyway.
Since we talk about *named* beans, like you have above with the `@Qualifier`, we have to use its opposite here as well:
```
/**
	 * The names of beans to check. The condition matches when each bean name specified is
	 * missing in the {@link ApplicationContext}.
	 * @return the name of beans to check
	 */
String[] name() default {};
```